hoomd/mpcd/CellList.cc

```diff
@@ -509,6 +526,22 @@ void mpcd::CellList::buildCellList()
             h_embed_cell_ids->data[cur_p - N_mpcd] = bin_idx;
             }
 
+        // compute the contribution of the particle to cell velocity
+        double4 cell_vel = h_cell_vel.data[bin_idx];
+        cell_vel.x += mass_i * vel_i.x;
+        cell_vel.y += mass_i * vel_i.y;
+        cell_vel.z += mass_i * vel_i.z;
+        cell_vel.w += mass_i;
+        h_cell_vel.data[bin_idx] = make_double4(cell_vel.x, cell_vel.y, cell_vel.z, cell_vel.w);
+        // compute optional cell properties
+        if (m_flags[mpcd::detail::thermo_options::energy])
+            {
+            double3 cell_energy = h_cell_energy.data[bin_idx];
+            cell_energy.x
+                += 0.5 * mass_i * (vel_i.x * vel_i.x + vel_i.y * vel_i.y + vel_i.z * vel_i.z);
+            cell_energy.z += __int_as_double(1);
+            h_cell_energy.data[bin_idx] = make_double3(cell_energy.x, 0.0, cell_energy.z);
+            }
```

Comment:
This may not work right as written (or at least, I'm unsure) because `__int_as_double` stashes an int into the bits of a double, so I don't know what happens if you then try to add those bits. However, here, you only need to add up the energy, as you will compute the temperature and can write the number of particles later.
```suggestion
        if (m_flags[mpcd::detail::thermo_options::energy])
            {
            h_cell_energy.data[bin_idx].x
                += 0.5 * mass_i * (vel_i.x * vel_i.x + vel_i.y * vel_i.y + vel_i.z * vel_i.z);
            }
```

We may reconsider whether double3 is the right data structure for this, or if we should do something different, later.

Reply:
The energy.y is referred to in the `CellThermoCompute` as the degrees of freedom. I thought that since this data was being collected at all, that suggested that degrees of freedom was not always equal to the number of particles in the cell, even if in practice they always were and the code didn't currently allowed for a scenario where they didn't.

If degrees of freedom always equals number of particles in the cell, then there's not really a point in energy.y at all. We could store the energy data in a double2 or something.

Reply:
It was probably written this way to help coalesce the write transaction for the cell property and/or to help coalesce the read transaction for the thermostatting. If it's the first, there is no benefit to that now, and the second may only be negligible. Are energy and temperature ever needed together? Otherwise, I would suggest breaking them into separate `GPUArray<double>`. The number of particles from the current array should be fine.

Reply:
> Are energy and temperature ever needed together?

No. SRD only ever uses the temperature value.

Reply:
That's what I figured, then use, please break apart as suggested above.